File: library/src/main/java/no/wtw/android/architectureutils/externalpayment/ExternalPaymentWebViewActivity.java
package no.wtw.android.architectureutils.externalpayment;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.Build;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.Toast;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.ViewById;

import no.wtw.android.architectureutils.view.ProgressOverlayView;

@EActivity(resName = "external_payment_web_view_activity")
public abstract class ExternalPaymentWebViewActivity extends Activity implements ExternalPaymentWebViewListener {

    @ViewById(resName = "external_payment_web_view")
    protected WebView webView;
    @ViewById(resName = "progress_overlay")
    protected ProgressOverlayView progressOverlay;

    protected ExternalPaymentWebViewClient client;

    @SuppressLint("SetJavaScriptEnabled")
    @AfterViews
    protected void afterViews() {
        client = new ExternalPaymentWebViewClient(this);
        WebSettings webSettings = webView.getSettings();
        webSettings.setDomStorageEnabled(true);
        webSettings.setDatabaseEnabled(true);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
            webSettings.setDatabasePath("/data/data/" + webView.getContext().getPackageName() + "/databases/");
        }
        webSettings.setJavaScriptEnabled(true);
        webSettings.setJavaScriptCanOpenWindowsAutomatically(true);
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
            CookieManager.getInstance().setAcceptThirdPartyCookies(webView, true);
        webView.setWebViewClient(client);
        try {
            webView.loadUrl(getExternalPaymentUrl());
            webView.requestFocus();
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(this, e.getMessage(), Toast.LENGTH_SHORT).show();
            finish();
        }
    }

    @Override
    public void setIsLoading(boolean isLoading) {
        if (isLoading)
            progressOverlay.setVisibility(View.VISIBLE);
        else
            progressOverlay.setVisibility(View.GONE);
    }

    @Override
    public void onExternalPaymentCancelled() {
        setResult(RESULT_CANCELED);
        finish();
    }

    @Override
    public String getExternalPaymentCancelledUrl() {
        return "mobillett:transcancel";
    }

    @Override
    public String getExternalPaymentSuccessUrl() {
        return "mobillett:transsuccess";
    }

    protected abstract String getExternalPaymentUrl() throws Exception;

}

File: library/src/main/java/no/wtw/android/architectureutils/externalpayment/ExternalPaymentWebViewActivity.kt
package no.wtw.android.architectureutils.externalpayment

import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.view.View.GONE
import android.view.View.VISIBLE
import android.viewbinding.library.activity.viewBinding
import android.webkit.CookieManager
import android.webkit.WebView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import no.wtw.android.architectureutils.databinding.ExternalPaymentWebViewActivityBinding
import no.wtw.android.architectureutils.view.ProgressOverlayView

abstract class ExternalPaymentWebViewActivity :
    AppCompatActivity(),
    ExternalPaymentWebViewListener {

    private val viewBinding by viewBinding<ExternalPaymentWebViewActivityBinding>()
    private lateinit var client: ExternalPaymentWebViewClient

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)

        ViewCompat.setOnApplyWindowInsetsListener(viewBinding.root) { _, windowInsets ->
            val insets =
                windowInsets.getInsets(WindowInsetsCompat.Type.ime() or WindowInsetsCompat.Type.statusBars() or WindowInsetsCompat.Type.tappableElement() or WindowInsetsCompat.Type.displayCutout())
            viewBinding.root.setPadding(insets.left, insets.top, insets.right, insets.bottom)
            WindowInsetsCompat.CONSUMED
        }

        client = ExternalPaymentWebViewClient(this)
        getWebView().settings.apply {
            domStorageEnabled = true
            databaseEnabled = true
            @SuppressLint("SetJavaScriptEnabled")
            javaScriptEnabled = true
            javaScriptCanOpenWindowsAutomatically = true
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
            CookieManager.getInstance().setAcceptThirdPartyCookies(getWebView(), true)
        getWebView().webViewClient = client
        try {
            getWebView().loadUrl(getExternalPaymentUrl())
            getWebView().requestFocus()
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(this@ExternalPaymentWebViewActivity, e.message, Toast.LENGTH_SHORT).show()
            finish()
        }
    }

    override fun setIsLoading(isLoading: Boolean) {
        getWebView().visibility = if (isLoading) GONE else VISIBLE
        getProgressOverlayView().visibility = if (isLoading) VISIBLE else GONE
    }

    override fun onExternalPaymentCancelled() {
        setResult(RESULT_CANCELED)
        finish()
    }

    override fun getExternalPaymentCancelledUrl() = "mobillett:transcancel"
    override fun getExternalPaymentSuccessUrl() = "mobillett:transsuccess"
    override fun isExternalPaymentSuccessUrlExactMatch() = false
    override fun isExternalPaymentCancelledUrlExactMatch() = false
    protected abstract fun getExternalPaymentUrl(): String

    // Either use the default viewBinding, or supply your own AND override these methods
    protected open fun getWebView(): WebView = viewBinding.externalPaymentWebView
    protected open fun getProgressOverlayView(): ProgressOverlayView = viewBinding.progressOverlay

}